File: Programmers/[42891] 무지의 먹방 라이브/KingDonggyu/42891.js
function solution(food_times, k) {
  const foodCount = food_times.length;
  const foods = food_times
    .map((time, i) => ({ number: i + 1, time: time }))
    .sort((a, b) => a.time - b.time);

  let prevTime = 0;

  for (let i = 0; i < foodCount; i++) {
    const leftoverCount = foodCount - i;
    const eatTime = (foods[i].time - prevTime) * leftoverCount;

    prevTime = foods[i].time;

    if (k < eatTime) {
      const leftoverFoods = foods.slice(i).sort((a, b) => a.number - b.number);
      return leftoverFoods[k % leftoverCount].number;
    }

    k -= eatTime;
  }

  return -1;
}

File: Programmers/[42891] 무지의 먹방 라이브/SUbbb/42891.java
import java.util.*;

class Food implements Comparable<Food> {
    int time;
    int idx;
    
    public Food(int time, int idx) {
        this.time = time;
        this.idx = idx;
    }
    
    @Override
    public int compareTo(Food f) {
        return this.time - f.time;
    }
}

class Solution {
    public int solution(int[] food_times, long k) {
        Queue<Food> pq = new PriorityQueue<>();
        // 모든 음식을 먹는데 걸리는 총 시간
        long foodSum = 0;
        int len = food_times.length;
        
        for (int i = 0; i < len; i++) {
            pq.add(new Food(food_times[i], i));
            foodSum += food_times[i];
        }
    
        // 음식을 다 섭취했는데 아직 K초가 되지 않았다면 -1 반환
        if (foodSum <= k) return -1;
        
        // 먹는데 사용한 시간
        long totalTime = 0;
        // 직전에 다 먹은 음식 시간
        long prevTime = 0;
        
        // 현재 음식을 다 먹을 수 있는 경우, totalTime을 증가
        while (totalTime + ((pq.peek().time - prevTime) * len) <= k) {
            int now = pq.poll().time;
            totalTime += (now - prevTime) * len;
            len -= 1;
            prevTime = now;
        }
        
        ArrayList<Food> remainFoods = new ArrayList<>();
        
        while (!pq.isEmpty()) remainFoods.add(pq.poll());
        
        // idx를 기준으로 오름차순 정렬
        remainFoods.sort(Comparator.comparingInt(o -> o.idx));
        
        return remainFoods.get((int) ((k - totalTime) % len)).idx + 1;
    }
}